File: go/odoid/decode.go
package odoid

import (
	"fmt"
	"strings"
)

// Decode decodes an OdoID string back to its originating integer.
//
// The input is uppercased before lookup, so lowercase letters that are valid
// in the charset (e.g. "0a0000") are accepted. The excluded characters I, L,
// and O remain invalid even after uppercasing.
//
// Returns an error if id is empty, has unsupported length, or contains
// characters absent from the positional charset.
func Decode(id string) (uint64, error) {
	if id == "" {
		return 0, fmt.Errorf("OdoID must be a non-empty string")
	}

	upper := strings.ToUpper(id)
	length := len(upper)

	if err := assertLength(length); err != nil {
		return 0, err
	}

	var n uint64

	for i := 0; i < length; i++ {
		charset := getCharset(i)
		base := uint64(len(charset))
		v := strings.IndexByte(charset, upper[i])

		if v < 0 {
			return 0, &InvalidCharacterError{Char: upper[i], Position: i + 1}
		}

		n = n*base + uint64(v)
	}

	return n, nil
}

File: lua/odoid/decode.lua
--- OdoID decoding — string to integer.

local charsets = require("odoid.charsets")
local errors   = require("odoid.errors")
local enc      = require("odoid.encode")

local M = {}

--- Decodes an OdoID string back to its originating integer.
---
--- The input is uppercased before lookup, so lowercase letters that are valid
--- in the charset (e.g. "0a0000") are accepted. The excluded characters
--- I, L, and O remain invalid even after uppercasing.
---
--- @param  id string   OdoID string (6, 7, or 8 characters).
--- @return number|nil, table|nil  n on success, or nil + error table on failure.
function M.decode(id)
  if not id or id == "" then
    return nil, { type = "EmptyInputError", message = "OdoID must be a non-empty string" }
  end

  local upper  = id:upper()
  local length = #upper

  local err = enc.assert_length(length)
  if err then return nil, err end

  local n = 0
  for i = 1, length do
    local charset = charsets.get_charset(i - 1)  -- 0-based
    local base    = #charset
    local ch      = upper:sub(i, i)
    local v       = charset:find(ch, 1, true)     -- plain find, 1-based
    if not v then
      return nil, errors.invalid_character(ch, i)
    end
    n = n * base + (v - 1)
  end

  return n, nil
end

return M
